bundles/org.openhab.binding.bmwconnecteddrive/src/main/java/org/openhab/binding/bmwconnecteddrive/internal/discovery/VehicleDiscovery.java

```diff
@@ -0,0 +1,227 @@
+    private final Logger logger = LoggerFactory.getLogger(VehicleDiscovery.class);
+    private static final int DISCOVERY_TIMEOUT = 10;
+    private Optional<ConnectedDriveBridgeHandler> bridgeHandler = Optional.empty();
+
+    public VehicleDiscovery() {
+        super(SUPPORTED_THING_SET, DISCOVERY_TIMEOUT, false);
+    }
+
+    public void onResponse(VehiclesContainer container) {
+        List<Vehicle> vehicles = container.vehicles;
+        vehicles.forEach(vehicle -> {
+            if (bridgeHandler.isPresent()) {
+                ThingUID bridgeUID = bridgeHandler.get().getThing().getUID();
+                // the DriveTrain field in the delivered json is defining the Vehicle Type
+                String vehicleType = vehicle.driveTrain.toLowerCase();
+                SUPPORTED_THING_SET.forEach(entry -> {
+                    if (entry.getId().equals(vehicleType)) {
+                        ThingUID uid = new ThingUID(entry, vehicle.vin, bridgeUID.getId());
+                        Map<String, String> properties = new HashMap<>();
+                        // Dealer
+                        if (vehicle.dealer != null) {
+                            properties.put("dealer", vehicle.dealer.name);
+                            properties.put("dealerAddress", vehicle.dealer.street + " " + vehicle.dealer.country + " "
+                                    + vehicle.dealer.postalCode + " " + vehicle.dealer.city);
+                            properties.put("dealerPhone", vehicle.dealer.phone);
+                        }
+
+                        // Services & Support
+                        properties.put("servicesActivated", getObject(vehicle, Constants.ACTIVATED));
+                        String servicesSupported = getObject(vehicle, Constants.SUPPORTED);
+                        String servicesNotSupported = getObject(vehicle, Constants.NOT_SUPPORTED);
+                        if (vehicle.statisticsAvailable) {
+                            servicesSupported += Constants.STATISTICS;
+                        } else {
+                            servicesNotSupported += Constants.STATISTICS;
+                        }
+                        properties.put(Constants.SERVICES_SUPPORTED, servicesSupported);
+                        properties.put("servicesNotSupported", servicesNotSupported);
+                        properties.put("supportBreakdownNumber", vehicle.breakdownNumber);
+
+                        // Vehicle Properties
+                        if (vehicle.supportedChargingModes != null) {
+                            StringBuilder chargingModes = new StringBuilder();
+                            vehicle.supportedChargingModes.forEach(e -> {
+                                chargingModes.append(e).append(Constants.SPACE);
+                            });
+                            properties.put("vehicleChargeModes", chargingModes.toString());
```

Comment:
I assume that the extra space at the end isn't intentional.
```suggestion
                            properties.put("vehicleChargeModes", String.join(Constants.SPACE, vehicle.supportedChargingModes));
```

Reply:
fixed